cips/cip-draft_ica.md

```diff
@@ -0,0 +1,921 @@
+  app.IBCKeeper.ChannelKeeper, // may be replaced with middleware such as ics29 fee
+  app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
+  app.AccountKeeper, scopedICAHostKeeper, app.MsgServiceRouter(),
+)
+
+// Create Interchain Accounts AppModule
+// Since only the host module is registered, nil is passed as the controller keeper
+icaModule := ica.NewAppModule(nil, &app.ICAHostKeeper)
+
+// Create a host IBC module 
+icaHostIBCModule := icahost.NewIBCModule(app.ICAHostKeeper)
+
+// Register host route
+ibcRouter.
+  AddRoute(icahosttypes.SubModuleName, icaHostIBCModule)
+...
+
+// Register Interchain Accounts AppModule
+app.moduleManager = module.NewManager(
+  ...
+  icaModule,
+)
+
+...
+
+// Add Interchain Accounts to begin blocker logic
+app.moduleManager.SetOrderBeginBlockers(
+  ...
+  icatypes.ModuleName,
+  ...
+)
+
+// Add Interchain Accounts to end blocker logic
+app.moduleManager.SetOrderEndBlockers(
+  ...
+  icatypes.ModuleName,
+  ...
+)
+
+// Add Interchain Accounts module InitGenesis logic
+app.moduleManager.SetOrderInitGenesis(
+  ...
+  icatypes.ModuleName,
+  ...
+)
+
+// initParamsKeeper init params keeper and its subspaces
+func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino, key, tkey sdk.StoreKey) paramskeeper.Keeper {
+  ...
+  paramsKeeper.Subspace(icahosttypes.SubModuleName)
+  ...
+}
+```
+
+## Rationale
+
+- Permissionless: An interchain account may be created by any actor without the approval of a third party (e.g. chain governance). Note: Individual implementations may implement their own permissioning scheme, however the protocol must not require permissioning from a trusted party to be secure.
+- Fault isolation: A controller chain must not be able to control accounts registered by other controller chains. For example, in the case of a fork attack on a controller chain, only the interchain accounts registered by the forked chain will be vulnerable.
+- The ordering of transactions sent to an interchain account on a host chain must be maintained. Transactions must be executed by an interchain account in the order in which they are sent by the controller chain.
+- If a channel closes, the controller chain must be able to regain access to registered interchain accounts by simply opening a new channel.
+- Each interchain account is owned by a single account on the controller chain. Only the owner account on the controller chain is authorized to control the interchain account. The controller chain is responsible for enforcing this logic.
+- The controller chain must store the account address of any owned interchain accounts registered on host chains.
+- A host chain must have the ability to limit interchain account functionality on its chain as necessary (e.g. a host chain can decide that interchain accounts registered on the host chain cannot take part in staking).
+
+## Backwards Compatibility
+
+No backward compatibility issues found
```

Comment:
This is a state machine breaking change. There is no backwards compatibility. The feature must be introduced in a major version only

Reply:
updated